File: klu erp try/gateway-node/src/services/search/brave.provider.js
/**
 * Brave Search API Provider
 * Integration with Brave Web Search REST API.
 */

const config = require('../../config');
const logger = require('../../utils/logger');
const { sanitizeText, sanitizeUrl } = require('../../utils/sanitizer');

class BraveSearchProvider {
  constructor() {
    this.name = 'brave';
    this.baseUrl = 'https://api.search.brave.com/res/v1/web/search';
  }

  /**
   * Search Brave Web Search API
   * @param {string} query - Search query
   * @param {number} maxResults - Max results count
   * @returns {Promise<Array>} Array of results
   */
  async search(query, maxResults = 8) {
    if (!config.search.braveApiKey) {
      logger.debug('Brave Search API key not configured, skipping');
      return [];
    }

    try {
      const url = `${this.baseUrl}?q=${encodeURIComponent(query)}&count=${maxResults}&text_decorations=false&safesearch=moderate`;
      const response = await fetch(url, {
        headers: {
          Accept: 'application/json',
          'X-Subscription-Token': config.search.braveApiKey,
        },
        signal: AbortSignal.timeout(config.search.timeoutMs),
      });

      if (!response.ok) {
        logger.warn(`Brave Search API returned HTTP ${response.status}`);
        return [];
      }

      const data = await response.json();
      const results = [];

      if (data.web && Array.isArray(data.web.results)) {
        for (const item of data.web.results) {
          results.push({
            title: sanitizeText(item.title),
            snippet: sanitizeText(item.description),
            source_url: sanitizeUrl(item.url),
            thumbnail_url: sanitizeUrl(item.thumbnail?.src || ''),
          });
        }
      }

      return results;
    } catch (err) {
      logger.warn(`Brave Search API error: ${err.message}`);
      return [];
    }
  }
}

module.exports = new BraveSearchProvider();

File: klu erp try/gateway-python/app/services/search/brave_provider.py
"""
Brave Search Web API Provider
Fetches live web search results using official Brave Search REST API.
"""

import httpx
from typing import List, Dict, Any

from app.config import settings
from app.utils.logger import logger
from app.utils.sanitizer import sanitize_text, sanitize_url

BRAVE_SEARCH_ENDPOINT = "https://api.search.brave.com/res/v1/web/search"


class BraveSearchProvider:
    def __init__(self):
        self.api_key = settings.brave_api_key

    async def search(self, query: str, limit: int = 8) -> List[Dict[str, Any]]:
        if not self.api_key:
            return []

        url = f"{BRAVE_SEARCH_ENDPOINT}?q={httpx.URL(query)}&count={limit}"
        headers = {
            "Accept": "application/json",
            "Accept-Encoding": "gzip",
            "X-Subscription-Token": self.api_key,
        }

        async with httpx.AsyncClient(timeout=3.5) as client:
            resp = await client.get(url, headers=headers)
            if resp.status_code != 200:
                raise Exception(f"Brave Search returned status {resp.status_code}")

            data = resp.json()
            web_results = data.get("web", {}).get("results", [])

            results = []
            for item in web_results[:limit]:
                results.append({
                    "title": sanitize_text(item.get("title", "")),
                    "snippet": sanitize_text(item.get("description", "")),
                    "source_url": sanitize_url(item.get("url", "")),
                    "thumbnail_url": sanitize_url(item.get("thumbnail", {}).get("src", "")),
                })

            return results


brave_provider = BraveSearchProvider()
